### PR TITLE
Trello-166 R&M CRSNOTE notification requests are idempotent

### PR DESCRIPTION
**Pull request by mahalrmoj**
https://trello.com/c/dbIMCNNo/166-make-notification-requests-idempotent

**What does this pull request do?**
The notification service creates contacts in nDelius for the event specified in the request. This change makes that request idempotent, using the following attributes to determine if the contact already exists for commission-rehabilitation-services:

- Offender Id
- NSI Id
- Contact Type (In this case CRSNOTE)
- Contact Date
- Contact Time

This key may not be applicable to all client contexts. This key is defined in the code to apply to CRS Contacts.

**What is the intent behind these changes?**
The intent is to make the link between R&M and NDelius more robust and remove the opportunity for duplicates to arise.

**Any breaking changes?**
There are no breaking API changes as such. However the call is now idempotent.

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/delius/jpa/standard/repository/ContactRepository.java [hunk 1]
```diff
@@ -48,4 +48,17 @@
     Optional<Contact> findByContactIdAndOffenderIdAndContactTypeAttendanceContactIsTrueAndSoftDeletedIsFalse(Long contactId, Long offenderId);
 
     Optional<Contact> findByContactIdAndOffenderIdAndSoftDeletedIsFalse(Long contactId, Long offenderId);
+
+    @Query("SELECT contact FROM Contact contact "
```
Comment (electrified) [inline src/main/java/uk/gov/justice/digital/delius/jpa/standard/repository/ContactRepository.java:52]: This multi line string could be a Text Block

File: src/main/java/uk/gov/justice/digital/delius/service/NotificationService.java [hunk 2]
```diff
@@ -33,10 +42,16 @@ public NotificationResponse notifyContact(final String crn, final Long sentenceI
 
         assertAppointmentType(request.getContactType());
 
-        final var newContact = makeNewContact(crn, sentenceId, request);
-        final var contactDto = deliusApiClient.createNewContact(newContact);
-
-        return NotificationResponse.builder().contactId(contactDto.getId()).build();
+        return matchExistingContact(crn, request)
+            .map(contactId -> {
+                log.debug(format("Notification already exists for CRN: %s, Sentence: %d, ContactType: %s and ContactDateTime: %s", crn, sentenceId, request.getContactType(), request.getContactDateTime()));
```
Comment (electrified) [inline src/main/java/uk/gov/justice/digital/delius/service/NotificationService.java:47]: This log message doesn't include everything used by the DB search. e.g. should it log NSI ID too?

File: src/main/java/uk/gov/justice/digital/delius/service/NotificationService.java [hunk 3]
```diff
@@ -59,6 +74,22 @@ private void assertAppointmentType(String contactTypeCode) {
         }
     }
 
+    private Optional<Long> matchExistingContact(String crn, NotificationCreateRequest request) {
+        Offender offender = offenderRepository.findByCrn(crn).orElseThrow(() -> new BadRequestException(format("Cannot find offender with CRN: %s", crn)));
+        var contactLocalDate = toLondonLocalDate(request.getContactDateTime());
+        var contactStartLocalTime = toLondonLocalTime(request.getContactDateTime()).truncatedTo(ChronoUnit.SECONDS);
+
+        return contactRepository.findByOffenderAndNsiIdAndContactTypeAndContactDateAndSoftDeletedIsFalse(
```
Comment (electrified) [inline src/main/java/uk/gov/justice/digital/delius/service/NotificationService.java:82]: Should this match on the requirement ID as that is a parameter too?

What about the note text/ team / staff member?

File: src/main/java/uk/gov/justice/digital/delius/service/NotificationService.java (hunk 2 above)
Comment (mahalrmoj) [reply src/main/java/uk/gov/justice/digital/delius/service/NotificationService.java:47]: Yeah will log it and perhaps as a WARN to so we can see how often it arises

File: src/main/java/uk/gov/justice/digital/delius/service/NotificationService.java (hunk 3 above)
Comment (mahalrmoj) [reply src/main/java/uk/gov/justice/digital/delius/service/NotificationService.java:82]: Will discuss with @mxco86 but the basic reasoning was that within an NSI there should never be a duplicate R&M event for the offender at any point in time.

Comment (mahalrmoj) [reply src/main/java/uk/gov/justice/digital/delius/service/NotificationService.java:82]: Conclusion is that we are going to make the method CRS specific and in that context the current equality check is adequate. The naming will make it clear.

**Review by electrified — APPROVED** [review]

**Pull request merged**